File: LeetCodeProblemsGo/LeetCode3479FruitsIntoBasketsIII/LeetCode3479FruitsIntoBasketsIII.go
package LeetCode3479FruitsIntoBasketsIII

func buildTree(segmentTree []int, baskets []int, index, start, end int) {
	if start == end {
		segmentTree[index] = baskets[start]
		return
	}
	mid := (start + end) >> 1
	buildTree(segmentTree, baskets, (index<<1)+1, start, mid)
	buildTree(segmentTree, baskets, (index<<1)+2, mid+1, end)
	segmentTree[index] = max(segmentTree[(index<<1)+1], segmentTree[(index<<1)+2])
}
func updateNode(segmetTree []int, index int) {
	if index == 0 {
		return
	}
	parent := (index - 1) >> 1
	segmetTree[parent] = max(segmetTree[(parent<<1)+1], segmetTree[(parent<<1)+2])
	updateNode(segmetTree, parent)
}
func fillBasket(segmentTree []int, index int) {
	segmentTree[index] = 0
	updateNode(segmentTree, index)
}
func query(segmentTree []int, fruit, index, start, end int) {
	if start == end {
		fillBasket(segmentTree, index)
	}
	if fruit > segmentTree[index] {
		return
	}
	mid := (start + end) >> 1

	if fruit <= segmentTree[(index<<1)+1] {
		query(segmentTree, fruit, (index<<1)+1, start, mid)
		return
	}
	if fruit <= segmentTree[(index<<1)+2] {
		query(segmentTree, fruit, (index<<1)+2, mid+1, end)
		return
	}

}
func numOfUnplacedFruits(fruits []int, baskets []int) int {
	length := len(baskets)
	segmentTree := make([]int, length<<2)
	buildTree(segmentTree, baskets, 0, 0, length-1)
	notPlacedFruits := 0
	for _, fruit := range fruits {
		if fruit > segmentTree[0] {
			notPlacedFruits++
			continue
		}
		query(segmentTree, fruit, 0, 0, length-1)
	}
	return notPlacedFruits
}

File: LeetCodeProblemsCpp/LeetCodeProblemsCpp/Problems/LeetCode3479FruitsIntoBasketsIII.cpp
#include "LeetCode3479FruitsIntoBasketsIII.h"

#include <iostream>
using  namespace std;
void buildTree(vector<int>& segmentTree,vector<int>& baskets,int index,int start,int end)
{
    if (start==end)
    {
        segmentTree[index]=baskets[start];
        return;
    }
    int mid=(start+end)>>1;
    buildTree(segmentTree,baskets,index*2+1,start,mid);
    buildTree(segmentTree,baskets,index*2+2,mid+1,end);
    segmentTree[index]= max (segmentTree[index*2+1],segmentTree[index*2+2]) ;
}
void UpdateElement(vector<int>& segmentTree,int index)
{
    if (index==0)
        return;
    int parentIndex=(index-1)>>1;
    segmentTree[parentIndex]=max(segmentTree[parentIndex*2+1],segmentTree[parentIndex*2+2]) ;
    UpdateElement(segmentTree,parentIndex);
}
void FillBasket(vector<int>& segmentTree,int index)
{
    segmentTree[index]=-1;
    UpdateElement(segmentTree,index);
}
void queryFruit(vector<int> & segmentTree,int index,int start,int end,int fruit)
{
    if (fruit>segmentTree[index])
        return;
    if (start==end)
    {
        FillBasket(segmentTree,index);
        return;
    }
    int mid=(start+end)>>1;
    if (fruit<=segmentTree[(index<<1)+1])
    {
        queryFruit(segmentTree,(index<<1)+1,start,mid,fruit);
        return;
    }
    if (fruit<=segmentTree[(index<<1)+2])
    {
        queryFruit(segmentTree,(index<<1)+2,mid+1,end,fruit);
        return;
    }
}
int LeetCode3479FruitsIntoBasketsIII::numOfUnplacedFruits( vector<int>& fruits,  vector<int>& baskets)
{
    int length=static_cast<int> (baskets.size());
    vector<int> segmentTree(length * 4,-1);
    buildTree(segmentTree,baskets,0,0,length-1);
    int unplacedFruits=0;
    for (int i=0;i<length;i++)
    {
        if (fruits[i]>segmentTree[0])
        {
            unplacedFruits++;
            continue;
        }
        queryFruit(segmentTree,0,0,length-1,fruits[i]);
    }
    return unplacedFruits;
}

void LeetCode3479FruitsIntoBasketsIII::Run()
{
    vector<int> fruits{59,59};
    vector<int> baskets{27,46};
    cout<<numOfUnplacedFruits(fruits,baskets)<<"\n";
}
